pkg/dockerregistry/auth/openshift.go

```diff
@@ -0,0 +1,159 @@
+package auth
+
+import (
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"net/http"
+	"strings"
+
+	log "github.com/Sirupsen/logrus"
+	ctxu "github.com/docker/distribution/context"
+	registryauth "github.com/docker/distribution/registry/auth"
+	authorizationapi "github.com/openshift/origin/pkg/authorization/api"
+	"github.com/openshift/origin/pkg/dockerregistry"
+	"golang.org/x/net/context"
+)
+
+func init() {
+	registryauth.Register("openshift", registryauth.InitFunc(newAccessController))
+}
+
+type contextKey int
+
+var bearerTokenKey contextKey = 0
+
+func WithBearerToken(parent context.Context, bearerToken string) context.Context {
+	return context.WithValue(parent, bearerTokenKey, bearerToken)
+}
+
+func BearerTokenFrom(ctx context.Context) (string, bool) {
+	bearerToken, ok := ctx.Value(bearerTokenKey).(string)
+	return bearerToken, ok
+}
+
+type AccessController struct {
+}
+
+var _ registryauth.AccessController = &AccessController{}
+
+type authChallenge struct {
+	err error
+}
+
+var _ registryauth.Challenge = &authChallenge{}
+
+// Errors used and exported by this package.
+var (
+	ErrTokenRequired          = errors.New("authorization header with basic token required")
+	ErrTokenInvalid           = errors.New("failed to decode basic token")
+	ErrOpenShiftTokenRequired = errors.New("expected openshift bearer token as password for basic token to registry")
+	ErrNamespaceRequired      = errors.New("repository namespace required")
+	ErrOpenShiftAccessDenied  = errors.New("openshift access denied")
+)
+
+func newAccessController(options map[string]interface{}) (registryauth.AccessController, error) {
+	fmt.Println("Using OpenShift Auth handler")
+	return &AccessController{}, nil
+}
+
+// Error returns the internal error string for this authChallenge.
+func (ac *authChallenge) Error() string {
+	return ac.err.Error()
+}
+
+// ServeHttp handles writing the challenge response
+// by setting the challenge header and status code.
+func (ac *authChallenge) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	// WWW-Authenticate response challenge header.
+	// See https://tools.ietf.org/html/rfc6750#section-3
+	w.Header().Add("WWW-Authenticate", fmt.Sprintf("Basic realm=openshift"))
```

Comment:
Sorry to make you change this again - I just read what Docker is doing. See e.g. https://github.com/docker/distribution/blob/master/registry/auth/token/accesscontroller.go#L106. Basically, always print `Basic realm=<realm>`, and optionally append `,error=<error string>` if necessary.

I also think the realm should be configurable, defaulting to `openshift` if nothing is specified.